Feature/build enhancements (#27)
redhat-marketplace/redhat-marketplace-operator: docs/release-process.md

> 

Reviewer: From line 61, should Bugfixes start off of develop?
```suggestion
Hotfixes are started off of master. Bugfixes off of develop.
```

Author: Hotfix is a fix against live production code. The master branch is always the latest available production source. 

A bugfix is a fix that can wait to be completed - it's not a high severity. So you treat it like a feature. It's not a feature thought because you're fixing an issue.

  A new branch called release/x.x.x will be made for you and pushed to the repository.

1. Operator images should be built and pushed for you with the build. Additionally there will be a github check that will create assets for you. You will need to publish the images in partner connect.
  1. Verify the checks for the PR are passing.
  1. Use the "stable-" assets from Generate Bundle workflow check to update partner connect. The bundle is a zip file with a zip inside of it. You will want to unzip the top file and submit the zip inside of it. Name will start with rhm-op-bundle.

>   1. You can use the `test-catalog-source` to install to your local CRC or to another openshfit to install the latest operator bundles for verification. You can just `kube apply -f` the file.

Reviewer: Small typo. 
```suggestion
  1. You can use the `test-catalog-source` to install to your local CRC or to another openshfit to test the latest operator bundles for verification. You can just `kube apply -f` the file.
```

1. Upload your generated bundle to [partner connect](#updating-partner-connect). And publish it when it passes. More details are provided in the [updating partner connect section](#updating-partner-connect).

1. Once the release is finished. Submit a PR to merge to master. Master build will deploy images, make the final bundle for upload to update stable.

1. Submit PR to merge master back into develop. Approve and merge into develop.

### Hotfix

Same as release, but change git flow commands to `git flow hotfix` and start from master branch.

## Manual Releases

### Prerequesites

To do all the manual actions. You will need to have these additional local commands.

1. [operator-sdk](https://sdk.operatorframework.io/docs/install-operator-sdk/)
1. [operator-courier](https://github.com/operator-framework/operator-courier)

### Release or Bugfix

*Warning*: to do these steps you need pull/push access to master/develop branch. And admin access to the docker registry. Only to be done if github actions are down.

> 1. Start from 'develop' branch.

Reviewer: Numbering may work differently with code-blocks, it shows multiple comments starting at "1.", instead of incrementing to 2, 3 ... . I'll simply add a comment to each line where this shows up.

> 1. Run

Reviewer: Numbering not incrementing properly. (1. 2. 3. ...)

  ```sh
  git flow release start $(make current-version)
  ```

  A new branch called release/x.x.x will be made for you and pushed to the repository.

> 1. Generate the csv files and commit them. The release branch should be used to create manifests for the beta channel. Updates to the bundle in Partner connect will only impact beta.

Reviewer: Numbering not incrementing properly. (1. 2. 3. ...)

  ```sh
  make generate-csv
  git add ./deploy/olm-catalog
  ```

> 1. Now you need to bundle the images for the release

Reviewer: Numbering not incrementing properly. (1. 2. 3. ...)

  ```sh
  export IMAGE_REGISTRY=quay.io/rh-marketplace
  export OPERATOR_IMAGE_TAG=$(make current-version)

  make build push publish-image bundle
  ```

> 1. Upload your bundle to [partner connect](#updating-partner-connect). And publish it when it passes. The bundle command will leave a zip file for upload in the `./bundle`

Reviewer: Numbering not incrementing properly. (1. 2. 3. ...)

1. Publish your images in Partner Connect and wait for the Operator scans to pass.

1. If the operator scan fails, see the [troubleshooting section](#troubleshooting). Otherwise finish the release.

1. Once the release is finished. Run these commands:

  ```sh
  git flow release finish $(make current-version)
  ```

> 1. Create the release using the published tag. Notify #rhm-general channel.

Reviewer: Numbering not incrementing properly. (1. 2. 3. ...)

  ![Publish Image](./images/pc-publish-operator.png)
1. You'll need to make sure the image is published and all the scorecard tests pass.

> 1. If all the tests are passed and you're read, select publish and your new version of the operator will be released.

Reviewer: small typo 
```suggestion
1. If all the tests are passed and you're ready, select publish and the new version of the operator will be released.
```

### Troubleshooting

> If your upload fails it's scan here are some common problems:

Reviewer: Small typo
```suggestion
If your upload fails its scan, here are some common problems:
```

#### Image verification step fails

This is the last step in the process, normally it'll say quay.io/xx/xx@sha256... cannot be found. This normally means that the image you used was published in Partner Connect. Go back to the [Updating Partner Connect](#updating-partner-connect) and retry the first few steps. Investigating the image's sha may also be required. You can use the [skopeo]() tool to inspect images for their digests running `skopeo inspect docker://${YOUR_IMAGE}`.

#### Scorecard step fails

We run the scorecard as well in our Deploy [workflow](#workflows). You'll want to verify that the results of our scorecards tests are valid as well. There can be some false-positive results so a particular test failing does not always mean it's a bad manifest. If the same test is failing, fixing it for our runs will likely fix it for the manifest scanning on Partner Connect.

> ##### A CR test isn't running properly

Reviewer: Heading size is different 
```suggestion
#### A CR test isn't running properly
```

Sometimes if the example CRs in 'deploy/crds' is updated to the same name as other resources we deploy then the test will fail. This failure normally comes through as `xx resource already exists`. Double check the names are example- or otherwise unique.

## Workflows

> 

Reviewer: Potentially add a one liner describing what a Workflow is, for clarity.

Each workflow perform specific tasks on different branches. Together they help build and deliver software.

| Workflow  | Git Action  | Branches  | Description  |
|:--|:--|:--|:--|
| Bump Version  | Push  | develop, hotfix/\*\*  |  Bumps version when it detects a tag on develop that's already been created in git |
| Create Manifest  | Push  | release/\*\*, hotfix/\*\*  |  Creates the new version's manifest and commits it to the branch |

> | Deploy | Push  | master, develop, release/\*\*, hotfix/\*\*, feature/\*\*, bugfix/\*\*  |   |

Reviewer: Deploy is missing a description
